---
title: ROSALIND | Finding a Shared Motif (LCSM)
date: 2024-05-11 18:44:14
tags: rosalind
categories:
---

çµ¦å®šä¸€ FASTA æª”æ¡ˆï¼Œå…§å«æ•¸æ¢ç”± Aã€Cã€Gã€T çµ„æˆçš„å­—ä¸²ï¼Œæ±‚é€™äº›å­—ä¸²å…±äº«çš„æœ€é•·å­å­—ä¸²ã€‚

> A common substring of a collection of strings is a substring of every member of the collection. We say that a common substring is a longest common substring if there does not exist a longer common substring. For example, "CG" is a common substring of "ACGTACGT" and "AACCGTATA", but it is not as long as possible; in this case, "CGTA" is a longest common substring of "ACGTACGT" and "AACCGTATA".
> Note that the longest common substring is not necessarily unique; for a simple example, "AA" and "CC" are both longest common substrings of "AACC" and "CCAA".
>
> Given: A collection of k (kâ‰¤100) DNA strings of length at most 1 kbp each in FASTA format.
>
> Return: A longest common substring of the collection. (If multiple solutions exist, you may return any single solution.)

(https://rosalind.info/problems/lcsm/)

<!--more-->

# çŸ¥è­˜é»
å¦‚æœæˆ‘å€‘èƒ½å¤ é€éç†è«–è¨ˆç®—æˆ–æ˜¯è³‡æ–™åº«æ•¸æ“šå¾—çŸ¥æŸç¨® motif çš„çµ„æˆï¼Œä¾¿èƒ½æŒ‰ç…§ [Finding a motif in dna](https://5uperb0y.com/finding-a-motif-in-dna/) é€™é¡Œçš„è¦æ±‚ï¼Œåœ¨ä¸€æ¢ DNA ä¸­å°‹æ‰¾å·²çŸ¥ motif çš„ä½ç½®ã€‚ç„¶è€Œï¼Œä¸¦éæ‰€æœ‰ç ”ç©¶å°è±¡éƒ½èƒ½äº‹å‰å¾—çŸ¥ motif çš„è³‡è¨Šï¼Œæ‰€ä»¥æ›´å¸¸è¦‹çš„æ–¹æ³•æ˜¯é€éæ¯”è¼ƒå¤šçµ„åŸºå› é«”ï¼Œæ­¸ç´å‡ºåè¦†å‡ºç¾ä¸”åºåˆ—ç›¸ä¼¼çš„ motifã€‚

ç‚ºäº†ç°¡åŒ–å•é¡Œï¼Œé€™é¡Œåªè¦æ±‚æ‰¾å‡ºæœ€é•·çš„ motifï¼Œå› ç‚ºé•·åº¦æ„ˆé•·ï¼Œmotif æ¶‰åŠçš„åŠŸèƒ½å¾€å¾€è¶Šå¤šã€‚æ­¤å¤–ï¼Œé¡Œç›®é‚„å‡è¨­åŒä¸€ç¨® motif çš„åºåˆ—å®Œå…¨ä¸€è‡´ï¼Œé€™æ„å‘³è‘—é€™äº› motif åœ¨æ¼”åŒ–ä¸­æ²’ç™¼ç”Ÿä»»ä½•çªè®Šï¼Œæš—ç¤ºå…¶åŠŸèƒ½ç›¸ç•¶é‡è¦ï¼Œå…·æœ‰ç›¸ç•¶çš„ç©©å®šæ€§ã€‚æ›å¥è©±èªªï¼Œé€™äº› motif å¯èƒ½åæ˜ äº†å„åŸºå› é«”å…±äº«çš„æ ¸å¿ƒåŠŸèƒ½ã€‚

# é¡Œè§£

é€™é¡Œç›¸ç•¶æ–¼å°‹æ‰¾å„å­—ä¸²çš„æœ€é•·å…±åŒå­å­—ä¸²ï¼ˆlongest common substring, LCSï¼‰ã€‚æœ€ç›´è§€çš„è§£æ³•åœ¨æ–¼äº†è§£ã€Œ LCS çš„é•·åº¦ä¸å¤§æ–¼æœ€çŸ­å­—ä¸²çš„é•·åº¦ã€ï¼Œå› æ­¤å¯ä»¥åˆ—å‡ºæœ€çŸ­å­—ä¸²çš„æ‰€æœ‰å­å­—ä¸²ï¼Œå†æ‹¿é€™äº›å­å­—ä¸²ç”±é•·è‡³çŸ­ä¸€ä¸€èˆ‡å…¶å®ƒå­—ä¸²æ¯”è¼ƒã€‚ç¬¬ä¸€å€‹å‡ºç¾åœ¨æ‰€æœ‰å­—ä¸²çš„å­å­—ä¸²å°±æ˜¯é¡Œç›®è¦æ±‚çš„ LCSã€‚

ä»¥ä¸‹ä½¿ç”¨ python å¯¦ä½œé€™å€‹ç®—æ³•ï¼Œé¦–å…ˆæ‰¾åˆ°æœ€çŸ­çš„å­å­—ä¸² `min_s`ï¼Œæ±‚å‡ºå®ƒæ‰€æœ‰çš„å­å­—ä¸² `kmer`ï¼Œæœ€å¾Œç”±é•·è‡³çŸ­æª¢æŸ¥å­å­—ä¸²æ˜¯å¦å‡ºç¾åœ¨æ‰€æœ‰å­—ä¸²ä¸­ï¼Œä¸€æ—¦ç¬¦åˆæ¢ä»¶ï¼Œä¾¿ç›´æ¥å›å‚³çµæœã€‚ï¼ˆç”±æ–¼é€™è£¡æ˜¯ç”±çŸ­è‡³é•·ç”Ÿæˆ `kmer`ï¼Œæ‰€ä»¥å¾Œé¢æŠŠå®ƒå€’ç½® `kmer[::-1]` ä¾†æ”¹è®Šæ¯”å°çš„é †åºã€‚ï¼‰

```python
def lcs(strs):
    """Finding the longest common substring from multiple strings
    """
    min_s = min(strs, key = len)
    kmer = [
        min_s[i:i+k+1]
        for k in range(len(min_s) + 1)
        for i in range(len(min_s) - k)
    ]
    for mer in kmer[::-1]:
        if all(mer in s for s in strs):
            return mer
        else:
            pass
    return ""
```

ç”±æ–¼é¡Œç›®è¦æ±‚è¼¸å…¥ FASTA æª”æ¡ˆï¼Œæ‰€ä»¥éœ€è¦ä¸€å€‹è®€æª”çš„å‰¯ç¨‹å¼ã€‚FASTA æª”æ¡ˆæ˜¯ç´€éŒ„åºåˆ—çš„é€šç”¨æ–‡å­—æ ¼å¼ï¼Œ`>` é–‹é ­çš„åˆ—ç´€éŒ„ DNAã€RNAæˆ–è›‹ç™½è³ªåºåˆ—çš„åç¨±ï¼Œç›´åˆ°ç¢°åˆ°ä¸‹ä¸€å€‹ `>` ç¬¦è™Ÿä¹‹å‰çš„åˆ—å‰‡è¨˜éŒ„äº†åºåˆ—æœ¬èº«ã€‚ç‚ºäº†æ–¹ä¾¿åœ¨çµ‚ç«¯æ©Ÿç”¨è‚‰çœ¼åˆ¤è®€ï¼Œæœ‰äº› FASTA æª”æ¡ˆä¸­çš„åºåˆ—æœƒåˆ†æˆæ•¸åˆ—è¡¨ç¤ºã€‚

```
>Read_id
ACTGACTGATCGGATAGTAGTGATAGCGATCGATCAGCATATCGACTATCAGCTAGCTACAGCTGAGCATCGATCG
ACGATTACGACGATCAG
```

ä»¥ä¸‹ç¨‹å¼å°‡ FASTA è®€é€² python dictionaryã€‚å‡¡æ˜¯ `>` é–‹é ­çš„åˆ—ï¼Œå°±æŠŠå®ƒç•¶ä½œ dictionary çš„ keyï¼Œè€Œå…¶é¤˜çš„åˆ—å‰‡ä¸²æ¥èµ·ä¾†æˆç‚º dictionary çš„ valueã€‚

```python
def read_fasta(path):
    """Read a fasta file into a dictionary
    """
    with open(path, "r") as f:
        reads = {}
        for line in f:
            if line.startswith(">"):
                key = line[1:].strip()
                reads[key] = ""
            else:
                reads[key] = reads[key] + line.strip()
        return reads
```

çµåˆè®€æª”èˆ‡ LCS ä¸»ç¨‹å¼ï¼Œä¾¿èƒ½æ»¿è¶³é¡Œç›®çš„è¦æ±‚äº†ã€‚

```python
def lcsm(fasta):
    """Finding the longest common nucleotide sequence from multiple sequences in a fasta file
    """
    seqs = list(read_fasta(fasta).values())
    return(lcs(seqs))
```

# è¨è«–

## longest common substring & longest common subsequence

[Longest common substring] å’Œ [longest common subsequence] æ˜¯ä¸åŒçš„å•é¡Œï¼š"AB" æ˜¯ "ABXC" å’Œ "ABOC" çš„ longest common substringï¼Œ"ABC" å‰‡å¦ï¼›`[A, B, C]` æ˜¯ `[A, B, X, C]` å’Œ `[A, B, O, C]` çš„ longest common subsequenceï¼Œ`[A, B]` å‰‡å¦ã€‚

å› ç‚ºåœ¨ç”Ÿç‰©è³‡è¨Šçš„èªå¢ƒè£¡ï¼Œåºåˆ—ï¼ˆsequenceï¼‰ä¹Ÿç”¨æ–¼æè¿° DNA/RNA/è›‹ç™½è³ªï¼Œè€Œé€™äº›åºåˆ—è³‡è¨Šå¾€å¾€ä¹Ÿç”¨å­—ä¸²ï¼ˆstringï¼‰ç´€éŒ„ã€‚å› æ­¤ï¼Œç•¶åˆæƒ³é€™é¡Œæœ‰ä»€éº¼å·§å¦™è§£æ³•æ™‚ï¼Œèª¤å…¥æ­§é€”å»æŸ¥äº† longest common "subsequence" çš„è§£æ³•ã€‚

å¯«å‡ºçš„ç¨‹å¼æ²’æœ‰è¿”å›é æœŸçµæœæ™‚ï¼Œå•å• chatgpt æ€éº¼è§£ longest common subsequence é‚„è¦ºå¾—æ˜¯ä¸æ˜¯ AI è®Šç¬¨äº†ï¼Œæ€éº¼çµ¦ä½  Rosalind é¡Œç›®çš„å®Œæ•´è¼¸å‡ºå…¥çµæœé‚„å˜´ç¡¬ã€‚åˆ†æ¸…æ¥šé€™å…©å€‹å•é¡Œå¾Œï¼Œæˆ‘æ‰æ˜ç™½å°ä¸‘æ˜¯æˆ‘è‡ªå·±ğŸ¤¡ã€‚

ï¼ˆå¦‚æœæƒ³è¦æ‰¾åˆ°æ›´å¿«çš„ç®—æ³•ï¼Œå¯ä»¥é»é€² Longest common substring é€£çµçœ‹çœ‹ç¶­åŸºç™¾ç§‘çš„è§£é‡‹ï¼‰

[Longest common substring]: https://en.wikipedia.org/wiki/Longest_common_substring
[longest common subsequence]: https://en.wikipedia.org/wiki/Longest_common_subsequence

## æ‰€æœ‰å­—ä¸²çš„ LCS ä¸è¦‹å¾—èˆ‡ä¸€å°å­—ä¸²çš„ LCS ä¸€è‡´

æ˜ç™½å°ä¸‘æ˜¯æˆ‘è‡ªå·±ä¹‹å¾Œï¼Œæˆ‘è¬„éŒ„äº†ä¸€ä»½åˆ©ç”¨å‹•æ…‹è¦åŠƒå°‹æ‰¾å…©å­—ä¸² longest common substring çš„æ–¹æ³•ã€‚ç°¡è¨€ä¹‹ï¼Œé€™æ–¹æ³•æœƒä¸€ä¸€æ¯”å°å…©å­—ä¸²æ¯å€‹å­—ç¬¦ï¼Œä¸¦ä¸”å°‡æœ€é•·å­å­—ä¸²çš„é•·åº¦ç´€éŒ„åœ¨ä¸€å€‹çŸ©é™£ä¸­ã€‚å†æ¯”å°éæ‰€æœ‰å­—ç¬¦å¾Œï¼Œä¾¿èƒ½å¾—å‡º LCSã€‚

```python
def lcs(s1, s2):
    """Finding the longest common substring from two strings
    """
    dp = [[0] * (len(s2) + 1) for _ in range(len(s1) + 1)]
    max_len, end_pos = 0, 0
    for i in range(1, len(s1) + 1):
        for j in range(1, len(s2) + 1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = dp[i-1][j-1] + 1
                if max_len < dp[i][j]:
                    max_len = dp[i][j]
                    end_pos = i
                else:
                    pass
            else:
                dp[i][j] = 0
    return s1[end_pos - max_len:end_pos]
```

æˆ‘åŸå…ˆçš„ä»¥ç‚ºï¼Œä¸€å°å­—ä¸²çš„ LCS æœƒèˆ‡æ‰€æœ‰å­—ä¸²çš„ LCS ç›¸åŒï¼Œæ‰€ä»¥æˆ‘å¯ä»¥åƒæ˜¯ã€Œæ‰¾å‡ºæœ€å¤§çš„æ•¸å­—ã€ä¸€æ¨£ï¼Œæ‰¾å‡ºã€Œæœ€é•·çš„å…±åŒå­å­—ä¸²ã€ï¼Œ

```python
# é€™æ˜¯éŒ¯çš„!!!!!
def lcsm(fasta):
    """Finding the longest common nucleotide sequence from multiple sequences in a fasta file
    """
    seqs = list(read_fasta(fasta).values())
    m = seqs[0]
    for s in seqs[1:]:
        m = lcs(s, m)
    return m
```

é›–ç„¶é€™æ”¯ç¨‹å¼å¯é †åˆ©é€šé ROSALIND çš„æ¸¬è³‡ï¼Œä½†æˆ‘åœ¨è¨è«–å€ç«‹åˆ»çœ‹åˆ°ä¸€å‰‡åä¾‹ï¼Œæ‰ä¹–ä¹–å¯«æˆé¡Œè§£é‚£å€‹å½¢å¼ã€‚

> I'm not sure the LCS of all the strings would necessarily be a substring of the pairwise LCS's. For example, in this set of strings:
> AAAAATATATAACGT AAAAACCCCCACGT ACACACCCCCACGT
> The LCS of the first and second string will be AAAAA, the LCS of the second and third string will be CCCCC, but the LCS of all three strings will be ACGT.

