<!DOCTYPE html>
<html lang="zh-TW">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222"><meta name="generator" content="Hexo 6.3.0">

  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.2/css/all.min.css" integrity="sha256-XOqroi11tY4EFQMR9ZYwZWKj5ZXiftSx36RRuC3anlA=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancyapps-ui/5.0.31/fancybox/fancybox.css" integrity="sha256-gkQVf8UKZgQ0HyuxL/VnacadJ+D2Kox2TCEBuNQg5+w=" crossorigin="anonymous">

<script class="next-config" data-name="main" type="application/json">{"hostname":"5uperb0y.com","root":"/","images":"/images","scheme":"Gemini","darkmode":false,"version":"8.20.0","exturl":false,"sidebar":{"position":"left","width_expanded":320,"width_dual_column":240,"display":"post","padding":18,"offset":12},"copycode":{"enable":true,"style":null},"fold":{"enable":false,"height":500},"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":true,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":false,"async":false,"transition":{"menu_item":"fadeInDown","post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"搜尋...","empty":"我們無法找到任何有關 ${query} 的搜索結果","hits_time":"${hits} 找到 ${time} 個結果","hits":"找到 ${hits} 個結果"},"path":"/search.xml","localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false}}</script><script src="/js/config.js"></script>

    <meta name="description" content="hap.py 是比較 VCF 的工具，常配合標準資料集來評估變異分析的效度，它整合了資料處理、集合運算與指標統計等功能，方便用戶解決效度評估常見的問題。">
<meta property="og:type" content="blog">
<meta property="og:title" content="如何使用 hap.py 評估小片段變異分析的效度？">
<meta property="og:url" content="https://5uperb0y.com/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/index.html">
<meta property="og:site_name" content="5uperb0y&#39;s blog">
<meta property="og:description" content="hap.py 是比較 VCF 的工具，常配合標準資料集來評估變異分析的效度，它整合了資料處理、集合運算與指標統計等功能，方便用戶解決效度評估常見的問題。">
<meta property="og:locale" content="zh_TW">
<meta property="article:published_time" content="2024-11-05T13:30:51.000Z">
<meta property="article:modified_time" content="2024-11-05T13:31:01.182Z">
<meta property="article:author" content="5uperb0y (超級boy)">
<meta property="article:tag" content="bioinformatics">
<meta name="twitter:card" content="summary">


<link rel="canonical" href="https://5uperb0y.com/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-TW","comments":true,"permalink":"https://5uperb0y.com/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/","path":"the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/","title":"如何使用 hap.py 評估小片段變異分析的效度？"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>如何使用 hap.py 評估小片段變異分析的效度？ | 5uperb0y's blog</title>
  
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-X8W93VY06Z"></script>
  <script class="next-config" data-name="google_analytics" type="application/json">{"tracking_id":"G-X8W93VY06Z","only_pageview":false,"measure_protocol_api_secret":null}</script>
  <script src="/js/third-party/analytics/google-analytics.js"></script>







<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2443737670547895"
     crossorigin="anonymous"></script>
  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="headband"></div>

  <main class="main">
    <div class="column">
      <header class="header" itemscope itemtype="http://schema.org/WPHeader"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切換導航欄" role="button">
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">5uperb0y's blog</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger" aria-label="搜尋" role="button">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>



<nav class="site-nav">
  <ul class="main-menu menu"><li class="menu-item menu-item-home"><a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首頁</a></li><li class="menu-item menu-item-about"><a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>關於</a></li><li class="menu-item menu-item-tags"><a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>標籤<span class="badge">31</span></a></li><li class="menu-item menu-item-archives"><a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>歸檔<span class="badge">147</span></a></li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜尋
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup"><div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off" maxlength="80"
           placeholder="搜尋..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close" role="button">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div class="search-result-container no-result">
  <div class="search-result-icon">
    <i class="fa fa-spinner fa-pulse fa-5x"></i>
  </div>
</div>

    </div>
  </div>

</header>
        
  
  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目錄
        </li>
        <li class="sidebar-nav-overview">
          本站概要
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%80%A5%E7%94%A8%E7%9A%84%E6%99%82%E5%80%99%E8%AE%80%E9%80%99%E8%A3%A1"><span class="nav-number">1.</span> <span class="nav-text">急用的時候讀這裡</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%83%B3%E6%AF%94%E8%BC%83-VCFs-%E7%9A%84%E8%A9%B1%EF%BC%8C%E7%82%BA%E4%BB%80%E9%BA%BC%E4%B8%8D%E7%9B%B4%E6%8E%A5%E5%8F%96%E4%BA%A4%E9%9B%86%E5%92%8C%E5%B7%AE%E9%9B%86%EF%BC%9F"><span class="nav-number">2.</span> <span class="nav-text">想比較 VCFs 的話，為什麼不直接取交集和差集？</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#multi-allelic-vs-bi-allelic-variants"><span class="nav-number">2.1.</span> <span class="nav-text">multi-allelic vs. bi-allelic variants</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#multiple-SNVs-vs-single-MNV"><span class="nav-number">2.2.</span> <span class="nav-text">multiple SNVs vs. single MNV</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#left-aligned-Indels-vs-right-aligned-Indels"><span class="nav-number">2.3.</span> <span class="nav-text">left-aligned Indels vs. right-aligned Indels</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#multiple-short-Indels-vs-single-long-Indel"><span class="nav-number">2.4.</span> <span class="nav-text">multiple short Indels vs. single long Indel</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%85%B7%E9%AB%94%E4%BE%86%E8%AA%AA%EF%BC%8Chap-py-%E6%98%AF%E4%BB%80%E9%BA%BC%EF%BC%9F"><span class="nav-number">3.</span> <span class="nav-text">具體來說，hap.py 是什麼？</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E4%BD%BF%E7%94%A8-hap-py-%E9%9C%80%E8%A6%81%E6%BA%96%E5%82%99%E4%BB%80%E9%BA%BC%EF%BC%9F"><span class="nav-number">4.</span> <span class="nav-text">使用 hap.py 需要準備什麼？</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%A6%82%E4%BD%95%E8%A7%A3%E8%AE%80%E7%B5%90%E6%9E%9C%EF%BC%9F"><span class="nav-number">5.</span> <span class="nav-text">如何解讀結果？</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%95%B8%E6%93%9A%E5%88%86%E5%B1%A4-Stratification"><span class="nav-number">5.1.</span> <span class="nav-text">數據分層 (Stratification)</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%B7%B7%E6%B7%86%E7%9F%A9%E9%99%A3%E9%A0%85%E7%9B%AE-Confusion-matrix-elements"><span class="nav-number">5.2.</span> <span class="nav-text">混淆矩陣項目 (Confusion matrix elements)</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%95%88%E5%BA%A6%E6%8C%87%E6%A8%99-Accuracy-metrics"><span class="nav-number">5.3.</span> <span class="nav-text">效度指標 (Accuracy metrics)</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%93%81%E7%AE%A1%E6%8C%87%E6%A8%99-Quality-metrics"><span class="nav-number">5.4.</span> <span class="nav-text">品管指標 (Quality metrics)</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%A6%82%E6%9E%9C%E9%82%84%E6%9C%89%E8%88%88%E8%B6%A3%EF%BC%8C%E5%8F%AF%E4%BB%A5%E7%A0%94%E8%AE%80%E4%BB%A5%E4%B8%8B%E8%AB%96%E6%96%87%E8%88%87%E7%B0%A1%E4%BB%8B"><span class="nav-number">6.</span> <span class="nav-text">如果還有興趣，可以研讀以下論文與簡介</span></a></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">5uperb0y (超級boy)</p>
  <div class="site-description" itemprop="description">對世界豎起中指吧</div>
</div>
  <div class="links-of-author animated">
      <span class="links-of-author-item">
        <a href="https://github.com/5uperb0y" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;5uperb0y" rel="noopener me" target="_blank"><i class="fab fa-github fa-fw"></i></a>
      </span>
      <span class="links-of-author-item">
        <a href="/contact.5uperb0y.blog@gmail.com" title="E-Mail → contact.5uperb0y.blog@gmail.com" rel="noopener me"><i class="fa fa-envelope fa-fw"></i></a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.youtube.com/channel/UC0D5wTTzJD_fTp2_gVAzSTA" title="YouTube → https:&#x2F;&#x2F;www.youtube.com&#x2F;channel&#x2F;UC0D5wTTzJD_fTp2_gVAzSTA" rel="noopener me" target="_blank"><i class="fab fa-youtube fa-fw"></i></a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.instagram.com/superb0y_potato/" title="Instagram → https:&#x2F;&#x2F;www.instagram.com&#x2F;superb0y_potato&#x2F;" rel="noopener me" target="_blank"><i class="fab fa-instagram fa-fw"></i></a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.linkedin.com/in/5uperb0y/" title="Linkedin → https:&#x2F;&#x2F;www.linkedin.com&#x2F;in&#x2F;5uperb0y&#x2F;?" rel="noopener me" target="_blank"><i class="fab fa-linkedin fa-fw"></i></a>
      </span>
  </div>
  <div class="cc-license animated" itemprop="license">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh" class="cc-opacity" rel="noopener" target="_blank"><img src="https://cdnjs.cloudflare.com/ajax/libs/creativecommons-vocabulary/2020.11.3/assets/license_badges/small/by_nc_sa.svg" alt="Creative Commons"></a>
  </div>


<div class="links-of-blogroll motion-element links-of-blogroll-block">
<div class="links-of-blogroll-title">
    <!-- 設定你要的fa fa icon-->
    <i class="fa fa-history fa-" aria-hidden="true"></i>
    近期文章
</div>
<ul class="links-of-blogroll-list">
    <!---->
    
    
    <li class='pull-left' style="white-space:normal; text-align:left" >
      <time title="創建時間" itemprop="dateCreated datePublished" datetime="2024-11-05T21:30:51+08:00">
        11/05
      </time>
      <a href="/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/" title="如何使用 hap.py 評估小片段變異分析的效度？">如何使用 hap.py 評估小片段變異分析的效度？</a>
      <br>
      <br>
    </li>
    
    <li class='pull-left' style="white-space:normal; text-align:left" >
      <time title="創建時間" itemprop="dateCreated datePublished" datetime="2024-11-04T22:17:26+08:00">
        11/04
      </time>
      <a href="/wu-2021-the-vertical-society-of-good-girls/" title="《美好少女的垂直社會》">《美好少女的垂直社會》</a>
      <br>
      <br>
    </li>
    
    <li class='pull-left' style="white-space:normal; text-align:left" >
      <time title="創建時間" itemprop="dateCreated datePublished" datetime="2024-10-31T16:01:09+08:00">
        10/31
      </time>
      <a href="/tsukasa-2019-asu-no-sekai-de-hoshi-wa-kirameku-3/" title="《明日的世界將會星光燦爛 3 》">《明日的世界將會星光燦爛 3 》</a>
      <br>
      <br>
    </li>
    
    <li class='pull-left' style="white-space:normal; text-align:left" >
      <time title="創建時間" itemprop="dateCreated datePublished" datetime="2024-10-28T22:26:31+08:00">
        10/28
      </time>
      <a href="/nextflow-manage-input-with-json/" title="Nextflow｜讀取儲存於 JSON 的輸入資料">Nextflow｜讀取儲存於 JSON 的輸入資料</a>
      <br>
      <br>
    </li>
    
    <li class='pull-left' style="white-space:normal; text-align:left" >
      <time title="創建時間" itemprop="dateCreated datePublished" datetime="2024-10-27T16:45:49+08:00">
        10/27
      </time>
      <a href="/tsukasa-2019-asu-no-sekai-de-hoshi-wa-kirameku-2/" title="《明日的世界將會星光燦爛 2 》">《明日的世界將會星光燦爛 2 》</a>
      <br>
      <br>
    </li>
    
</ul>
</div>

        </div>
      </div>
    </div>

    
    <div class="sidebar-inner sidebar-blogroll">
      <div class="links-of-blogroll animated">
        <div class="links-of-blogroll-title"><i class="fa fa-globe fa-fw"></i>
          Links
        </div>
        <ul class="links-of-blogroll-list">
            <li class="links-of-blogroll-item">
              <a href="https://www.facebook.com/3minBioPhysics/" title="https:&#x2F;&#x2F;www.facebook.com&#x2F;3minBioPhysics&#x2F;" rel="noopener" target="_blank">三分鐘生物物理</a>
            </li>
            <li class="links-of-blogroll-item">
              <a href="https://yongfu.name/" title="https:&#x2F;&#x2F;yongfu.name&#x2F;" rel="noopener" target="_blank">Yongfu's Blog</a>
            </li>
            <li class="links-of-blogroll-item">
              <a href="https://wcw0310.github.io/" title="https:&#x2F;&#x2F;wcw0310.github.io&#x2F;" rel="noopener" target="_blank">Hi I'm WCW!</a>
            </li>
        </ul>
      </div>
    </div>
        <div class="pjax">
        <div class="sidebar-inner sidebar-post-related">
          <div class="animated">
              <div class="links-of-blogroll-title"><i class="fa fa-signs-post fa-fw"></i>
    相關文章
  </div>
  <ul class="popular-posts">
    <li class="popular-posts-item">
      <a class="popular-posts-link" href="/how-to-create-bwa-alt-files/" rel="bookmark">
        <time class="popular-posts-time">2024-10-14</time>
        <br>
      怎麼準備 bwa 所需的 .alt 檔？
      </a>
    </li>
  </ul>

          </div>
        </div>
        </div>
  </aside>


    </div>

    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-TW">
    <link itemprop="mainEntityOfPage" href="https://5uperb0y.com/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="5uperb0y (超級boy)">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="5uperb0y's blog">
      <meta itemprop="description" content="對世界豎起中指吧">
    </span>

    <span hidden itemprop="post" itemscope itemtype="http://schema.org/CreativeWork">
      <meta itemprop="name" content="如何使用 hap.py 評估小片段變異分析的效度？ | 5uperb0y's blog">
      <meta itemprop="description" content="">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          如何使用 hap.py 評估小片段變異分析的效度？
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">發表於</span>
      

      <time title="創建時間：2024-11-05 21:30:51 / 修改時間：21:31:01" itemprop="dateCreated datePublished" datetime="2024-11-05T21:30:51+08:00">2024-11-05</time>
    </span>

  
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody"><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/hint.css/2.4.1/hint.min.css"><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/hint.css/2.4.1/hint.min.css"><p><strong>hap.py</strong> 是比較 VCF 的工具，常配合標準資料集來評估變異分析的效度，它整合了資料處理、集合運算與指標統計等功能，方便用戶解決效度評估常見的問題。</p>
<span id="more"></span>

<h1 id="急用的時候讀這裡"><a href="#急用的時候讀這裡" class="headerlink" title="急用的時候讀這裡"></a>急用的時候讀這裡</h1><p>準備 <code>truth.vcf</code>、<code>query.vcf</code>、<code>reference.fa</code>、<code>reference.fa.fai</code>、<code>confident.bed</code> 放入工作目錄，下載並進入 hap.py 分析環境，執行腳本後檢查結果。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ docker pull quay.io/biocontainers/hap.py:0.3.15--py27hcb73b3d_0</span><br><span class="line">$ docker run -it -v &lt;/path/to/workdir&gt;:&lt;/path/to/workdir&gt; -w &lt;/path/to/workdir&gt; quay.io/biocontainers/hap.py:0.3.15--py27hcb73b3d_0 bash</span><br><span class="line">$ hap.py truth.vcf query.vcf -f confident.bed -o output_prefix -r reference.fa</span><br><span class="line">$ awk <span class="string">&#x27;&#123;print $1,$14&#125;&#x27;</span> output_prefix.summary</span><br><span class="line"></span><br><span class="line">Type  METRIC.F1_Score </span><br><span class="line"></span><br><span class="line">INDEL 0.5                ← 😫 結果很不好</span><br><span class="line"></span><br><span class="line">SNP   1                  ← 😁 結果非常好</span><br></pre></td></tr></table></figure>

<p>等待程式運算期間可以泡杯咖啡繼續讀，了解 hap.py 的用途與結果詮釋。</p>
<h1 id="想比較-VCFs-的話，為什麼不直接取交集和差集？"><a href="#想比較-VCFs-的話，為什麼不直接取交集和差集？" class="headerlink" title="想比較 VCFs 的話，為什麼不直接取交集和差集？"></a>想比較 VCFs 的話，為什麼不直接取交集和差集？</h1><p>因為相同變異可能有不同的表示方式，所以變異間非一一對應關係，若僅取交集和差集會錯估變異分析的效度，例如：</p>
<h2 id="multi-allelic-vs-bi-allelic-variants"><a href="#multi-allelic-vs-bi-allelic-variants" class="headerlink" title="multi-allelic vs. bi-allelic variants"></a>multi-allelic vs. bi-allelic variants</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	T	G,C	1000	PASS	.	GT	2/1</span><br></pre></td></tr></table></figure>
<p>or</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	T	C	1000	PASS	.	GT	0/1</span><br><span class="line">chr1	20000	.	T	G	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<h2 id="multiple-SNVs-vs-single-MNV"><a href="#multiple-SNVs-vs-single-MNV" class="headerlink" title="multiple SNVs vs. single MNV"></a>multiple SNVs vs. single MNV</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	TC	CA	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>
<p>or</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	T	C	1000	PASS	.	GT	0/1</span><br><span class="line">chr1	20001	.	C	A	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<h2 id="left-aligned-Indels-vs-right-aligned-Indels"><a href="#left-aligned-Indels-vs-right-aligned-Indels" class="headerlink" title="left-aligned Indels vs. right-aligned Indels"></a>left-aligned Indels vs. right-aligned Indels</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	20002	.	TCCT	TCT	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>
<p>or</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	20001	.	TC	T	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<h2 id="multiple-short-Indels-vs-single-long-Indel"><a href="#multiple-short-Indels-vs-single-long-Indel" class="headerlink" title="multiple short Indels vs. single long Indel"></a>multiple short Indels vs. single long Indel</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># ref:   TC-C-TG</span><br><span class="line"># input: TCTCCTG</span><br><span class="line">chr1	20001	.	C	CT	1000	PASS	.	GT	0/1</span><br><span class="line">chr1	20002	.	C	CC	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>
<p>or</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"># ref:   T--CCTG</span><br><span class="line"># input: TCTCCTG</span><br><span class="line">chr1	20000	.	T	TCT	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<h1 id="具體來說，hap-py-是什麼？"><a href="#具體來說，hap-py-是什麼？" class="headerlink" title="具體來說，hap.py 是什麼？"></a>具體來說，hap.py 是什麼？</h1><p>hap.py ≈ vcf preprocessor + comparison engine + metric summarizer</p>
<ul>
<li>vcf prprocessor：統一 VCF 格式，標準化變異表示法，方便比較引擎運作</li>
<li>comparison engine：考量變異表示法差異，正確統計混淆矩陣的各項元素（true positive, false positive, false negative）</li>
<li>metric sumarizer：依照混淆矩陣計算精準度、召回率和 F1 score 等效度指標。</li>
</ul>
<h1 id="使用-hap-py-需要準備什麼？"><a href="#使用-hap-py-需要準備什麼？" class="headerlink" title="使用 hap.py 需要準備什麼？"></a>使用 hap.py 需要準備什麼？</h1><ul>
<li><code>truth.vcf</code> 記錄了你認為可靠的變異，用以和 query.vcf 比較。經過 Genome in a Bottle (GIAB) Consortium 認證的變異資料集可在 <a target="_blank" rel="noopener" href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/">NCBI FTP 資料庫</a>下載。</li>
<li><code>query.vcf</code> 是你打算評估的 VCF 檔，標頭至少要含有 VCF 格式版本 (<code>##fileformat</code>)、contigs（<code>##contig=&lt;...&gt;</code>）、genotype（<code>##FORMAT=&lt;ID=GT,...&gt;</code>）及欄位名稱（<code>#CHROM ...</code>）。</li>
<li><code>confident.bed</code> 記錄了你要求 hap.py 檢查的基因體區域，這些區域以外的變異會標記為 UNK，不納入效度計算。”confident” 這名稱沿襲 GIAB 的慣例，位於 confident region 的變異經多項技術驗證，任何額外或遺漏的變異皆可視為誤報。</li>
<li><code>reference.fa</code> &amp; <code>reference.fa.fai</code> 是變異分析使用的參考基因體及其索引檔。<code>hap.py</code> 需要這些資訊來依據基因體區域計算效度，還有計算 het-hom 與 Ti&#x2F;Tv ratio 等品管指標。因此，要使用生成 truth.vcf 和 query.vcf 的參考基因體版本。</li>
<li><code>hap.py</code> &amp; its environment 可以參考<a target="_blank" rel="noopener" href="https://github.com/Illumina/hap.py?tab=readme-ov-file#installation">官方教學</a>安裝，或是使用 biocontainer 上的 <a target="_blank" rel="noopener" href="https://biocontainers.pro/tools/hap.py">image</a>。dockerhub 也有一份 <a target="_blank" rel="noopener" href="https://hub.docker.com/r/pkrusche/hap.py?ref=https://githubhelp.com">image</a> ，可是已數年沒有更新，所以內容落後 hap.py 的教學文件。</li>
</ul>
<p>執行分析前可以檢查一下 truth.vcf 和 query.vcf 的染色體表示法是否一致（<code>chrX</code> or <code>X</code>）、標頭內容和 INFO 欄位是否匹配，這些都是 hap.py 報錯的常見原因。</p>
<h1 id="如何解讀結果？"><a href="#如何解讀結果？" class="headerlink" title="如何解讀結果？"></a>如何解讀結果？</h1><p>hap.py 會在你使用 <code>-o</code> 指定的目錄輸出比對結果。精確度、召回率與 F1 score 等校度統計量紀錄在 <code>*.summary.csv</code>，其它次要的統計量則保留在 <code>*.extended.csv</code>。</p>
<p><code>*.summary.csv</code> 的欄位可分為數據分層和效度指標兩類，我習慣先看 Type 和 METRIC.F1_Score，只要 SNV 和 INDEL 的 F1 score 接近 1，那麼大家都會幸福。</p>
<table>
<thead>
<tr>
<th>Type</th>
<th>METRIC.F1_Score</th>
</tr>
</thead>
<tbody><tr>
<td>INDEL</td>
<td>0.99</td>
</tr>
<tr>
<td>SNP</td>
<td>0.99</td>
</tr>
</tbody></table>
<h2 id="數據分層-Stratification"><a href="#數據分層-Stratification" class="headerlink" title="數據分層 (Stratification)"></a>數據分層 (Stratification)</h2><p>用戶能要求 hap.py 獨立計算特定類別的變異之效度指標。因此，如果想更仔細檢驗分析結果，可以先查看數據分層的方式（詳情參考 <a target="_blank" rel="noopener" href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md#introduction">hap.py 的官方簡介</a>）。</p>
<ul>
<li><strong>Type</strong>：目前，hap.py 支援 SNV 和 INDEL 的效度評估，暫時沒有其它變異類型。</li>
<li><strong>Subtype</strong>：SNV 可細分為 transition （腺嘌呤與鳥糞嘌呤，胞嘧啶與胸腺嘧啶互換）和 transversion（嘌呤與嘧啶互換），INDEL 也有不同長度片段的插入與刪除。</li>
<li><strong>Filter</strong>：hap.py 能依據 VCF 檔 FILTER 欄位，獨立列出數值為 PASS 的變異之效度指標。比較合格與全數變異的效度指標，便能評估變異品管的成效。</li>
<li><strong>Subset</strong>：如果使用 <code>--stratification &lt;BED&gt;</code> 要求 hap.py 獨立計算特定區域的效度指標，就會多出 <em>Subset</em> 欄位註記分析結果所屬區域的名稱，這些名稱對應到 BED 檔的 ID 欄（第四欄）。</li>
</ul>
<h2 id="混淆矩陣項目-Confusion-matrix-elements"><a href="#混淆矩陣項目-Confusion-matrix-elements" class="headerlink" title="混淆矩陣項目 (Confusion matrix elements)"></a>混淆矩陣項目 (Confusion matrix elements)</h2><p>比對之前，hap.py 會先統一 <code>truth.vcf</code> 和 <code>query.vcf</code> 的格式（例如 #CHROM 是否包含 <code>chr</code>），再標準化 <code>query.vcf</code> 以利比較引擎運作。標準化的方式包含將 MNV 拆分為獨立的 SNV、拆分 multi-allelic 變異、依據基因型合併相同位置的變異等，詳見<a target="_blank" rel="noopener" href="https://github.com/Illumina/hap.py/blob/master/doc/normalisation.md">官方文件</a>）。</p>
<p>hap.py 預設的比較引擎為 <em>xmap</em>，它會考量變異表示法的差異，從而避免低估分析的正確性。假設 <code>truth.vcf</code> 包含一個 MNV，例如：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	TC	CA	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<p>標準化之後， <code>query.vcf</code> 會將 MNV 拆分為兩個獨立的 SNV。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">chr1	20000	.	T	C	1000	PASS	.	GT	0/1</span><br><span class="line">chr1	20001	.	C	A	1000	PASS	.	GT	0/1</span><br></pre></td></tr></table></figure>

<p>這兩組 VCF 檔實際上是同一種變異的不同表示方式。如果只取交集會得到空集合，但 <em>xmap</em> 能辨識這種狀況並正確標示出來。</p>
<ul>
<li><strong>TRUTH.TP</strong> 是出現於 <code>truth.vcf</code>，而且能在標準化 <code>query.vcf</code> 找到的變異數量。</li>
<li><strong>QUERY.TP</strong> 是出現於標準化 <code>query.vcf</code>，而且能在<code>truth.vcf</code> 找到的變異數量。</li>
</ul>
<p>TRUTH.TP 和 QUERY.TP 的數值有可能不一樣，這是因為在不同表示方式下，變異之間非一一對應。在上述案例中，<code>truth.vcf</code> 的一個 MNV 對應 <code>query.vcf</code> 的兩個 SNV，所以 TRUTH.TP 為 1，QUERY.TP 為 2。除了 MNV 和 SNV，序列比對以及基因型也跟變異的表示法有關。多種要素混合在一起就變得沒那麼直觀，所以我也還沒掌握所有案例。不過，如果對結果有疑慮都可以查看 <code>output_prefix.vcf.gz</code> 的 <em>TRUTH</em> 和 <em>QUERY</em> 欄位。它們註記了變異真偽的判定結果，可以依此推論 hap.py 標準化和比較變異的方式。</p>
<ul>
<li><strong>TRUTH.FN</strong>：<code>truth.vcf</code> 獨有的變異數量</li>
<li><strong>QUERY.FP</strong>：<code>query.vcf</code> 獨有的變異數量</li>
<li><strong>QUERY.UNK</strong> 是 <code>query.vcf</code> 位在 <em>confident.bed</em> 以外的變異數量。</li>
<li><strong>TRUTH.TOTAL</strong>：<em>truch.vcf</em> 的變異總數，相當於 <em>TRUTH.TP</em> + <em>TRUTH.FN</em>。</li>
<li><strong>QUERY.TOTAL</strong>：<code>query.vcf</code> 的變異總數，相當於 <em>QUERY.TP</em> + <em>QUERY.FP</em> + <em>QUERY.UNK</em>。</li>
</ul>
<p>hap.py 另外將 FP 分為兩類，但其實很少用到。</p>
<ul>
<li><strong>FP.gt</strong>：Genotype errors，相同位置 (<code>POS</code>)、相同變異 (<code>REF</code>&#x2F;<code>ALT</code>)、不同基因型 (<code>GT</code>)，例如 query GT &#x3D; 1&#x2F;1，truth GT &#x3D; 0&#x2F;1。</li>
<li><strong>FP.al</strong>：Allele errors，相同位置、不同變異、不同基因型。</li>
</ul>
<h2 id="效度指標-Accuracy-metrics"><a href="#效度指標-Accuracy-metrics" class="headerlink" title="效度指標 (Accuracy metrics)"></a>效度指標 (Accuracy metrics)</h2><ul>
<li><strong>METRIC.Recall</strong>：召回率，真實變異被檢出的比率，計算方式為 TRUTH.TP &#x2F; (TRUTH.TP + TRUTH.FN)。評估變異分析的效度時，通常會用多筆 <code>query.vcf</code> 對照單一 <code>truth.vcf</code>。召回率採計 TRUTH.TP 和 TRUTH.FN 以免數值因為 <code>query.vcf</code> 的表示法差異而浮動<sup id="fnref:1"><a href="#fn:1" rel="footnote"><span class="hint--top hint--error hint--medium hint--rounded hint--bounce" aria-label="假設兩組 `query.vcf` 用不同表示方式記錄相同的變異，它們照理要有相同的召回率，所以才會用不受 `query.vcf` 表示法影響的 `truth.vcf` 來計算。">1</span></a></sup>。</li>
<li><strong>METRIC.Precision</strong>：精準度，檢出變異為真實變異的比率，計算方式為 QUERY.TP &#x2F; (QUERY.TP + QUERY.FP)。</li>
<li><strong>METRIC.F1_Score</strong>：精準度與召回率的調和平均，計算方式為 2 * METRIC.Recall* METRIC.Precision &#x2F; (METRIC.Recall + METRIC.Precision)。</li>
<li><strong>METRIC.Frac_NA</strong>：QUERY.UNK &#x2F; QUERY.TOTAL。</li>
</ul>
<p>召回率高但精準度低的分析過度寬鬆，納入過多偽變異徒增診斷成本（全部個案都判陽性的檢驗有百分之百的召回率，卻一點用都沒有）；精準度高但召回率低的分析過度保守，遺漏過多真實變異限縮診斷效果（不會報錯卻驗不出真實變異的檢驗也沒什麼用）。F1 score 同時衡量了召回率和精準度，提供更符合直覺的效度估計。F1 score 使用調和平均的理由可以參考 <a target="_blank" rel="noopener" href="https://stackoverflow.com/questions/26355942/why-is-the-f-measure-a-harmonic-mean-and-not-an-arithmetic-mean-of-the-precision">why is the f measure a harmonic mean and not an arithmetic mean of the precision</a>。簡言之，調和平均對於差異懸殊數值的權重較低，所以唯有同時確保召回率和精準度，才能取得較高的 F1 score。</p>
<h2 id="品管指標-Quality-metrics"><a href="#品管指標-Quality-metrics" class="headerlink" title="品管指標 (Quality metrics)"></a>品管指標 (Quality metrics)</h2><p>Ti&#x2F;Tv ratio 和 het&#x2F;hom ratio 是評估變異分析品質的指標，和 <code>truth.vcf</code>&#x2F;<code>query.vcf</code> 的比較無關。這兩種指標在生物基因體內有一定範圍，而且其數值遠高於隨機值。通常，指標數值越高，變異分析的品質也越好。Ti&#x2F;Tv ratio 是 transition 和 transversion 兩種 SNV 數量的比值。假設鹼基發生 transition 和 transversion 的機率一致，那麼 Ti&#x2F;Tv ratio 應該要等於 0.5。因為 transition 有兩種變化途徑，但 transversions 有四種變化途徑，2 比 4 得 0.5。</p>
<p>依據轉譯密碼表，造成胺基酸變化的突變往往屬於 transversion，這種突變若發生在重要的蛋白質可能導致生物死亡。存在選擇壓力的情況下，生物體的 transversion 數量往往少於 transition，使得基因體的 Ti&#x2F;Tv ratio 遠高於 0.5。人類全基因體定序結果的 Ti&#x2F;Tv ratio 接近 2.1，全外顯子定序的 Ti&#x2F;Tv ratio 接近 3（因為 exon 比 intron 暴露到更多選擇壓力下）。假設定序或分析的錯誤皆為隨機發生，若全基因體變異分析的 Ti&#x2F;Tv ratio 比率遠低於 2.1，便可合理懷疑其中混入了太多隨機發生的錯誤，導致數值偏離預期。</p>
<p>het&#x2F;hom ratio (heterozygous&#x2F;homozygous variant ratio) 也是基於相似的原理。依據族群遺傳學的哈溫定律，het&#x2F;hom ratio 會等於 2，如果變異分析的結果偏離預期太多，也可能反映了某種錯誤存在。這兩種指標都只能用來評估一整批變異分析的結果，細節可參考 Guo et al. (2014). Three-stage quality control strategies for DNA re-sequencing data. Briefings in bioinformatics, 15(6), 879-889.</p>
<ul>
<li><strong>TRUE.TOTAL.TiTv_ratio</strong>：truth.vcf 的 Ti&#x2F;Tv ratio。</li>
<li><strong>QUERY.TOTAL.TiTv_ratio</strong>：query.vcf 的 Ti&#x2F;Tv ratio。</li>
<li><strong>TRUTH.TOTAL.het_hom_ratio</strong>：truth.vcf 的 het&#x2F;hom ratio。</li>
<li><strong>QUERY.TOTAL.het_hom_ratio</strong>：query.vcf 的 het&#x2F;hom ratio。</li>
</ul>
<p>在有正確答案的情況下，直接比較 <code>truth.vcf</code> 和 <code>query.vcf</code> 更為直觀且具體。不過，Ti&#x2F;Tv ratio 和 het&#x2F;hom ratio 能作為額外的查核資訊。如果它們的數值和 F1 score 呈不同趨勢，能提醒我們細查資料，確認是否發生非預期的錯誤。</p>
<h1 id="如果還有興趣，可以研讀以下論文與簡介"><a href="#如果還有興趣，可以研讀以下論文與簡介" class="headerlink" title="如果還有興趣，可以研讀以下論文與簡介"></a>如果還有興趣，可以研讀以下論文與簡介</h1><ul>
<li><a target="_blank" rel="noopener" href="https://medium.com/truwl/demystifying-benchmarking-a-guide-to-germline-variant-calling-metrics-9d6eee4f8e80">Karl Sebby. (2021). Demystifying benchmarking: A guide to germline variant calling metrics.</a>：簡單明瞭的部落格文章，把變異分析效度評估解釋得很清楚。</li>
<li><a target="_blank" rel="noopener" href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6699627/">Krusche et al. (2019). Best practices for benchmarking germline small-variant calls in human genomes. Nature biotechnology, 37(5), 555-560.</a>：GA4GH Benchmarking Team 發表的效度分析指引，可以了解變異表示法相關的偏誤，以及各項指標的設計原則。</li>
<li><a target="_blank" rel="noopener" href="https://pmc.ncbi.nlm.nih.gov/articles/instance/6699627/bin/NIHMS1533783-supplement-1.pdf">Supplementary of Krusche et al. (2019)</a>：前一則論文的補充資料，更詳盡地解釋了指標的公式。</li>
<li><a target="_blank" rel="noopener" href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md">Illumina&#x2F;hap.py&#x2F;doc&#x2F;happy.md</a> 以及 <a target="_blank" rel="noopener" href="https://github.com/Illumina/hap.py">Illumina&#x2F;hap.py: Haplotype VCF comparison tools.</a>：hap.py 官方手冊，能了解程式具體做了那些處理。</li>
<li><a target="_blank" rel="noopener" href="https://medium.com/dnanexus/benchmarking-state-of-the-art-secondary-variant-calling-pipelines-5472ca6bace7">Yih-Chii Hwang. (2020). Benchmarking state-of-the-art secondary variant calling pipelines.</a>：說明如何進行效度分析的部落格文章。<div id="footnotes"><hr><div id="footnotelist"><ol style="list-style: none; padding-left: 0; margin-left: 40px"><li id="fn:1"><span style="display: inline-block; vertical-align: top; padding-right: 10px; margin-left: -40px">1.</span><span style="display: inline-block; vertical-align: top; margin-left: 10px;">假設兩組 <code>query.vcf</code> 用不同表示方式記錄相同的變異，它們照理要有相同的召回率，所以才會用不受 <code>query.vcf</code> 表示法影響的 <code>truth.vcf</code> 來計算。<a href="#fnref:1" rev="footnote">↩</a></span></li></ol></div></div></li>
</ul>

    </div>

    
    
    

    <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/bioinformatics/" rel="tag"># bioinformatics</a>
          </div>

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/wu-2021-the-vertical-society-of-good-girls/" rel="prev" title="《美好少女的垂直社會》">
                  <i class="fa fa-angle-left"></i> 《美好少女的垂直社會》
                </a>
            </div>
            <div class="post-nav-item">
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">

  <div class="copyright">
    &copy; 
    <span itemprop="copyrightYear">2024</span>
    <span class="with-love">
      <i class="fa fa-heart"></i>
    </span>
    <span class="author" itemprop="copyrightHolder">5uperb0y (超級boy)</span>
  </div>

    </div>
  </footer>

  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>
  <div class="sidebar-dimmer"></div>
  <div class="back-to-top" role="button" aria-label="回到頂端">
    <i class="fa fa-arrow-up fa-lg"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/animejs/3.2.1/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/next-theme-pjax/0.6.0/pjax.min.js" integrity="sha256-vxLn1tSKWD4dqbMRyv940UYw4sXgMtYcK6reefzZrao=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/fancyapps-ui/5.0.31/fancybox/fancybox.umd.js" integrity="sha256-a+H7FYzJv6oU2hfsfDGM2Ohw/cR9v+hPfxHCLdmCrE8=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/lozad.js/1.16.0/lozad.min.js" integrity="sha256-mOFREFhqmHeQbXpK2lp4nA3qooVgACfh88fpJftLBbc=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/sidebar.js"></script><script src="/js/next-boot.js"></script><script src="/js/pjax.js"></script>

  <script src="https://cdnjs.cloudflare.com/ajax/libs/hexo-generator-searchdb/1.4.1/search.js" integrity="sha256-1kfA5uHPf65M5cphT2dvymhkuyHPQp5A53EGZOnOLmc=" crossorigin="anonymous"></script>
<script src="/js/third-party/search/local-search.js"></script>




  <script src="/js/third-party/fancybox.js"></script>



  




  

  <script class="next-config" data-name="enableMath" type="application/json">false</script><script class="next-config" data-name="mathjax" type="application/json">{"enable":true,"tags":"none","js":{"url":"https://cdnjs.cloudflare.com/ajax/libs/mathjax/3.2.2/es5/tex-mml-chtml.js","integrity":"sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI="}}</script>
<script src="/js/third-party/math/mathjax.js"></script>


  <script src="https://cdnjs.cloudflare.com/ajax/libs/quicklink/2.3.0/quicklink.umd.js" integrity="sha256-yvJQOINiH9fWemHn0vCA5lsHWJaHs6/ZmO+1Ft04SvM=" crossorigin="anonymous"></script>
  <script class="next-config" data-name="quicklink" type="application/json">{"enable":true,"home":true,"archive":true,"delay":true,"timeout":3000,"priority":true,"url":"https://5uperb0y.com/the-hitchhikers-guide-to-benchmarking-small-variant-calling-with-happy/"}</script>
  <script src="/js/third-party/quicklink.js"></script>

</body>
</html>
